### PR TITLE
Improve SWA graph reuse, generic masks, and kernel robustness

### PR DESCRIPTION
**Pull request by ani300**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [x] feature
- [x] cleanup
- [ ] documentation

#### What this PR does:

Follow-up to #3405 for the Gemma 3/4 integration in torch-spyre/hf-adapters#496.

- replaces position-dependent Python geometry with one required runtime additive `attention_mask`
- removes `cache_seqlen`, `buffer_origin`, `valid_start`, `decode_mask`, and the internal `window_band_mask` op
- keeps both prefill and decode on fused sliding-window attention; the enabled SWA path does not route through SDPA
- reuses one decode graph while mask contents and tensor cache indices change
- adds a required static `is_causal` guarantee: causal square prefill retains narrow per-window reads, while generic masks scan the complete physical cache in bounded chunks
- supports masks with non-causal regions, including Gemma 4 bidirectional vision blocks
- fixes online softmax for fully masked `-inf` chunks and removes dead synthetic accumulator initialization
- fixes D=256, transposed-query, batched MHA/GQA, and related compiler-pass issues exposed by the kernel
- adds a measured tiling selector for Gemma 3/4 production SWA shapes, while retaining the proven conservative fallback for uncalibrated shapes and hardware

The public schema is now:

```python
torch.ops.spyre.sliding_window_attention(
    query, key, value, attention_mask, window_size, is_causal, scale=None
)
```

`window_size`, `is_causal`, `scale`, shapes, and dtypes are static model configuration. Decode position, padding, valid rows, window placement, and bidirectional regions are tensor data. `is_causal=False` makes no semantic claim beyond disabling causal-only narrow reads; the tensor mask remains authoritative.

The SWA tiling selector uses only compile-time shape, dtype, core-count, and LX-budget facts, so it does not reintroduce specialization on mask contents or cache position. On the calibrated 32-core target it keeps all query heads in one coarse tile and selects 576-row KV blocks for Gemma 3 or 512-row KV blocks for Gemma 4. Other geometries keep the prior 512-row/at-most-four-head fallback exactly.

#### Which issue(s) this PR is related to:

- Follow-up to #3405
- Companion integration: torch-spyre/hf-adapters#496

#### Special notes for your reviewer:

Focused validation on current `main` (`769c9d2d`):

- selector/SDPA unit tests plus runtime-mask and non-causal regressions: 22 passed, 20 subtests passed
- Gemma 3 production compact-cache decode device test: passed
- Gemma 4 production decode and prefill device tests: 2 passed
- production selector: cosine 0.999998, no recompilation, expected 16-head/512-KV policy
- hf-adapters Gemma 4 Spyre A/B suite: 5 passed, including 70 decode steps across a cache roll
- hf-adapters focused CPU SWA suite: 36 passed
- Gemma 4 E2B CPU-vs-Spyre greedy-token comparison from the preceding revision: 5/5 top-1 agreement, no NaNs
- Ruff, formatting, py_compile, and diff checks pass

Measured warm E2E results from the SWA shape sweep:

| model / query rows | previous policy | selected policy | previous | selected | change |
|---|---|---|---:|---:|---:|
| Gemma 4 / 1 | H=4, KV=512 | H=16, KV=512 | 1.558 ms | 1.142 ms | -26.7% |
| Gemma 4 / 64 | H=4, KV=512 | H=16, KV=512 | 2.510 ms | 1.620 ms | -35.5% |
| Gemma 4 / 512 | H=4, KV=512 | H=16, KV=512 | 17.704 ms | 11.927 ms | -32.6% |
| Gemma 3 / 1 | H=4, KV=512 | H=8, KV=576 | 0.738 ms | 0.503 ms | -31.8% |
| Gemma 3 / 64 | H=4, KV=512 | H=8, KV=576 | 1.018 ms | 0.666 ms | -34.6% |

All successful sweep points met the correctness threshold and reported no recompilation. A final production-selector Gemma 4 decode rerun measured 1.156 ms E2E / 0.901 ms kernel with cosine 0.999998.

Explicit full-core work division was also tested rather than assumed. It is not currently valid for this decomposition: a query split propagates to the mask's singleton broadcast-head dimension, and KV splits encounter physical-stick placement constraints. The selected policies therefore improve the existing legal coarse tiling without hiding dropped or invalid work-division hints.

The compact 1088-column mask build plus host-to-device transfer measured 146.2 us median and is paid once per decode step, shared across all sliding layers. Generic non-causal prefill scans the full physical cache because future allowed regions are known only from runtime mask values; decode remains bounded by the compact cache.

#### Does this PR introduce a user-facing change?

Yes. This intentionally replaces the unshipped #3405 API. `attention_mask` is required and is the only runtime geometry input. `is_causal` is a required static access-plan guarantee, not position-dependent state.

### CODE REVIEW

**Review by dgrove-oss — DISMISSED** [review]
Summary

Follow-up to #3405 for the Gemma 3/4 integration: replaces sliding_window_attention's Python-side positional geometry (is_causal, cache_seqlen, buffer_origin, valid_start) with a single required runtime additive attention_mask tensor, enabling one compiled decode graph to be reused across changing cache positions; fixes an online-softmax NaN hazard for fully-masked KV chunks; and fixes a graph-editing operation-naming collision that can silently corrupt scheduler dependencies after a late pass removes operations.

Review

BLOCKERs

(none)

SUGGESTIONs

- _windowed_attention's full_cache branch reads the complete KV allocation in _SDPA_MAX_SEQUENCE_TILE_SIZE-sized chunks even for decode (Lq=1), which is correct but means a very long compact cache still pays a chunk count proportional to buffer_width, not the window. The PR body notes "use a compact cache for decode so this remains proportional to the sliding window" — worth double-checking the intended HF integration always allocates a compact (window-sized) cache for the runtime-mask decode path, since nothing in this op enforces that at the API boundary (a caller could pass a full-length KV allocation and silently get correct-but-slow behavior rather than a diagnostic).
- _to_cache_position_first (test helper, test_sliding_window_attention.py:285) constructs a SpyreTensorLayout by hand with an inline stride/shape derivation. Since this now backs three real regression tests (test_prefill_from_pinned_larger_cache_with_left_padding and two others) rather than being a one-off, a short comment on why this specific device_size/stride_map pairing matches "model cache update" layout would help a future reader confirm it's still correct if the real integration's layout ever shifts.

QUESTIONs

- _windowed_attention's KV-chunking loop is bounded by _SDPA_MAX_SEQUENCE_TILE_SIZE = 512 (an existing full-SDPA constant, reused here). Was this reused because it was already validated as a safe LX/matmul tile size at this head_dim range for SWA specifically, or is it a starting point that may need SWA-specific tuning later (e.g. via the same kind of cost-model approach as PR #4424)?

Verification performed

- Confirmed the online-softmax NaN fix numerically: independently reproduced the all-masked-leading-chunk scenario in Python (block_max clamp to torch.finfo(dtype).min before the max-reduction subtraction) and confirmed exp_scores for a fully-masked chunk contributes exactly zero, and that a later real chunk's correction term correctly zeroes out the garbage finite_min-seeded running state rather than producing NaN. This matches the new dedicated regression test test_runtime_mask_skips_fully_masked_leading_chunks.
- Confirmed register_operation_after_graph_edit's premise is a real hazard, not a hypothetical one: read upstream GraphLowering.register_operation's actual source and confirmed it unconditionally derives the next name from len(self.operations) with no collision check, so a late pass that shortens the operations list before registering a new op can silently collide with and overwrite an in-use name. The new helper's loop-until-unused-name fix is correct, and the new test test_late_operation_registration_does_not_reuse_a_removed_slot directly exercises the collision scenario. All three call sites (graph_editor.py, lx_context_switching.py ×2) were migrated.
- Confirmed no dangling references to the removed public surface (window_band_mask op, band_batch, band_valid_start, check_valid_start) remain anywhere in the repo outside the files this PR itself edits — every removed symbol's only other users were the test/decomposition files already updated in the same diff.
- Confirmed _SDPA_MAX_SEQUENCE_TILE_SIZE and _sdpa_num_head_tiles are pre-existing helpers already used elsewhere in decompositions.py, correctly reused rather than reinvented.
- Confirmed the new tests/inductor/test_pass_utils.py file carries the correct 14-line Apache 2.0 license header.
- StarDep/WeakDep fix in pass_utils.py's iteration_space/iteration_space_from_op: WeakDep is an ordering-only dependency edge with no .ranges in the general case, so including it in the loop-dimension derivation was a latent bug independent of this PR's main feature; the new test test_reduction_iteration_space_ignores_weak_dependencies directly targets it.
- CI: 128 pass / 20 pending / 1 skipping at last check, zero failures. DCO: 7/7 commits signed.
- Dual-path impact: compiled-path only (torch_spyre/_inductor/), consistent with test coverage added only under tests/inductor/.

Verdict

APPROVE

This is a well-tested, deliberate API redesign with strong regression coverage for the two real bugs it fixes (NaN on fully-masked chunks, operation-name collision after graph edits) and a sensible rationale for the breaking schema change, given the explicit companion-PR coordination named in the PR body. The two QUESTIONs are about cross-repo rollout coordination and tuning provenance, not correctness concerns.

**Comment by ani300** [comment]
Addressed all items from the dgrove-oss review:

- Documented exactly how the cache-position-first test layout maps logical B/H/L/D axes to device axes and strides.
- Documented that runtime-mask decode work scales with physical cache width and that callers should use a compact allocation.
- Verified and strengthened the companion hf-adapters integration: anchored decode now rejects a noncompact sliding cache for Gemma 3 and the shared Gemma 4 dense/MoE/VLM path instead of silently taking the correct-but-slow path (torch-spyre/hf-adapters#496, commit aef1559).
- Replaced the original generic 512-row assumption for calibrated Gemma shapes with the measured SWA tiling selector and sweep results in commit 6025fa4f.

Focused hf-adapters cache coverage: 18 passed. Static checks pass in both repositories.

**Review by dgrove-oss — APPROVED** [review]
Summary

Replaces sliding-window attention's Python-side geometry (cache_seqlen, buffer_origin, valid_start, decode_mask, window_band_mask) with a single required runtime additive attention_mask tensor, so position/padding/window changes never trigger recompilation. Causal square prefill keeps its narrow static KV-window read; every other shape (non-causal, decode, chunked prefill) now scans the full compact cache in bounded chunks driven by a measured tiling selector. Also fixes an online-softmax NaN bug (-inf - -inf) on fully-masked KV chunks, and fixes an unrelated late-graph-edit operation-naming collision bug in the scratchpad passes.

Review

BLOCKERs

None.

SUGGESTIONs

- The new attention_mask validation guards added to spyre_sliding_window_attention (decompositions.py, ~1466–1499: non-positive window_size, non-positive/mismatched seqlen_q/cache_capacity, non-stick-aligned cache_capacity, and mask shape/dtype/device mismatches) have no unit test exercising any of them — every existing test passes a well-formed mask. Worth a small parametrized test asserting each raises Unsupported.
- register_operation_after_graph_edit (pass_utils.py) and its use in graph_editor.py/lx_context_switching.py fix a real bug (name collision after a late pass shortens graph.operations) but are not mentioned in the PR description, which is scoped to SWA. Worth calling out explicitly in the PR body since it's a behavior change to shared scratchpad infrastructure, not SWA-specific.
- No docs/ update accompanies the public op schema change (new required attention_mask param; removed cache_seqlen/buffer_origin/valid_start/decode_mask). Since torch.ops.spyre.sliding_window_attention is consumed by hf-adapters rather than being a widely-documented user-facing op, this is minor, but a docstring-only contract (as currently done) means anyone grepping docs/ for this op's shape will find nothing.

QUESTIONs

- None outstanding — the new _select_swa_tiling selector's calibrated/fallback split is well-covered by test_swa_tiling.py, the NaN-chunk fix is covered by a targeted regression test (test_runtime_mask_skips_fully_masked_leading_chunks, asserting torch.isfinite), and the mask-slicing-under-read_start interaction with narrow causal windows is exercised by several existing prefill tests.


Verdict

APPROVE — the API simplification is sound, well-tested (causal/non-causal, decode/prefill/chunked-prefill, GQA, head_dim 128/256, graph-reuse-across-mask-values, and the NaN-chunk edge case are all covered), and the incidental scratchpad bugfix is minimal and unit-tested. The two SUGGESTIONs (missing validation tests, undisclosed scratchpad fix) are non-blocking.
